### PR TITLE
nonblocking concurrent task execution via background workers

### PR DESCRIPTION
**Pull request by aykut-bozkurt**
DESCRIPTION: Adds nonblocking concurrent task execution via background workers.

Improvement on our background task monitoring API (PR #6296) to support concurrent and nonblocking task execution.

Mainly we have a queue monitor background process which forks task executors for `Runnable` tasks and then monitors their status by fetching messages from shared memory queue in nonblocking way.

We can summarize Queue monitor loop steps:
1. Queue monitor tries to fetch and assign `Runnable` tasks to _task executor workers_ as far as max worker limits allow (it stores <task, handle-to-task-worker> pairs in the hash table),
2. Queue monitor iterates over tasks in the hash table and drains the output from shared memory queue which is set up to communicate with the task executor (for each iteration, each task starts at ExecutionStarted state and ends up at ExecutionEnd state),
3. If all tasks returned `EWOULDBLOCK`, then the monitor sleeps 1 sec to let them progress and generate some output so that we can prevent CPU load.

#### How to handle `Failed` task?
If a task fails, the monitor sets its state back to `Runnable` and backs off an amount of time specified by the retry count.

#### How to handle cancellation of a `Running` task?
If a `Running` task is concurrently cancelled, its status is changed to `Cancelling` by job cancel api call so that monitor can report the cancel message for the task before making its status as `Cancelled`.

### CODE REVIEW

**Comment by aykut-bozkurt** [comment]
- [ ] Prevent CPU load on EWOULD_BLOCK flood,

- [ ] Cancel message is missing when a concurrent cancel occurs. (missing the message due to nonblocking receive on queue)

File: src/backend/distributed/utils/background_jobs.c [hunk 1]
```diff
@@ -446,22 +473,40 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
-		BackgroundWorkerHandle *handle =
-			StartCitusBackgroundTaskExecuter(databaseName, userName, task->command,
-											 task->taskid, &seg);
+		StringInfo message = NULL;
+
+		bool handleEntryFound = false;
+		BackgroundExecutorHashEntry *handleEntry =
+			hash_search(backgroundExecutorHandles, (void *) &task->taskid, HASH_ENTER,
+						&handleEntryFound);
+		if (!handleEntryFound)
```
Comment (onderkalaci) [inline src/backend/distributed/utils/background_jobs.c:489]: Do we plan to have a configurable upper bound on the number of allowed concurrent executions? Probably yes, something like the executor pool size would be neat

Comment (onderkalaci) [reply src/backend/distributed/utils/background_jobs.c:489]: Note that, we are also discussing how to make shard moves parallel. But then, that decision is probably should be given by `rebalance_table_shards()`, and this bg worker should just pick the next avaliable task(s).

I'm planning to discuss this with @thanodnl and @JelteF today, so maybe decision coming from there could impact how the parallel bg worker works.

For now, I think you can focus on making things work in bg worker in parallel, what we plan to do is probably mostly a follow up task on top of this work anyway.

Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:502]: Whenever our monitoring bg detects a new ready task, we spawn a new task executor bg. I noticed that pooling is done at db level with `max_worker_processes` parameter which can only be set at server startup. With the default configuration, we can spawn 5 task executor bg. (other 3 is used by other Citus bgs) When I increase the  `max_worker_processes` param, we can create more parallel task executor bgs.

Comment (JelteF) [reply src/backend/distributed/utils/background_jobs.c:502]: While max_worker_processes is indeed the hard upper bound, we don't really want to hit it.  So, I agree with onder that it's good to have some upper bound on the task executor bg-workers. A separate GUC for this would be in line with how postgres manages everything that consumes from max_worker_processes.

**Comment by aykut-bozkurt** [comment]
> @microsoft-github-policy-service agree [company="{your company}"]

@microsoft-github-policy-service agree [company="citusdata"]

**Comment by aykut-bozkurt** [comment]
@microsoft-github-policy-service agree company="citusdata"

**Comment by aykut-bozkurt** [comment]
> @aykut-bozkurt please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Microsoft"

**Comment by aykut-bozkurt** [comment]
It could be nicer if we can further prune FSM states to make it more readable. Any advice is welcomed.

File: src/backend/distributed/metadata/metadata_utility.c [hunk 2]
```diff
@@ -3375,23 +3375,35 @@ BackgroundTaskReadyToRun(BackgroundTask *task)
...
+	 * Ordering of status is important.
+	 * 1) We should change Cancelling status to Cancelled,
+	 * 2) We should then select runnable or running tasks.
+	 */
 	BackgroundTaskStatus taskStatus[] = {
-		BACKGROUND_TASK_STATUS_RUNNABLE
+		BACKGROUND_TASK_STATUS_CANCELLING,
+		BACKGROUND_TASK_STATUS_RUNNABLE,
+		BACKGROUND_TASK_STATUS_RUNNING
 	};
```
Comment (thanodnl) [inline src/backend/distributed/metadata/metadata_utility.c:3402]: I am surprised by this change, wouldn't any task that is ready to be started at this stage be marked as `RUNNABLE`?

To me a task that is `CANCELLING` should not be touched, because the backend of the task itself is getting a signal. This should trigger the backend to stop execution and the state should transition not when we are looking for a next task, but instead when the stopped backend is inspected and the associated task updated.

Same would de for `RUNNING` these tasks are not ready to start because they already should be started.

I didn't look deeply into the rest of the change, but this part stood out to me while skimming the change.

Either we are doing funky things here, or the comment above is very misleading. And doesn't explain why we are iterating `CANCELLING` and `RUNNING` tasks.

Comment (aykut-bozkurt) [reply src/backend/distributed/metadata/metadata_utility.c]: You are right. Here some updates. I refactored the code such that queue monitor looks for runnable tasks and updates their states in FSM loop. It looks for task table to fetch runnable tasks and also has a hash table to store currently running task and executor pairs. Cancelling is also updated in the FSM now. I made FSM states little bit more readable.

File: src/backend/distributed/utils/background_jobs.c [hunk 3]
```diff
@@ -231,7 +240,6 @@ citus_job_wait_internal(int64 jobid, BackgroundJobStatus *desiredStatus)
 		}
 
 		/* sleep for a while, before rechecking the job status */
-		CHECK_FOR_INTERRUPTS();
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:234]: Hmm, I assumed we will only cancel a task by using citus_job_cancel API call. Then, we should check interrupts there to allow ctrl+c. But anyway, when you cancel a query which waits for a task execution, doesn't the cancellation signal direct to psql backend process? How do we handle another process' signal on queue monitor process? I lack info for that part.

File: src/backend/distributed/utils/background_jobs.c [hunk 4]
```diff
@@ -385,292 +648,339 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
+
+					case TaskConcurrentCancelCheck:
+					{
+						/*
+						 * here we take exclusive lock on pg_dist_background_task table to prevent a
+						 * concurrent modification. We first check if a task is concurrently cancelled
+						 * or deleted. Accordingly, we decide to make the task status as cancelled or running.
+						 */
+
+						StartTransactionCommand();
```
Comment (marcocitus) [inline src/backend/distributed/utils/background_jobs.c:763]: I find it slightly hard to know for sure that we eventually commit this transaction in this loop. Is there a different way to do this? e.g. we always open a transaction?

File: src/backend/distributed/utils/background_jobs.c [hunk 5]
```diff
@@ -385,292 +648,339 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
+						}
+						else if (mq_res == SHM_MQ_WOULD_BLOCK)
+						{
+							/*
+							 * still running the task but had no output for current iteration.
+							 * commit current transaction and jump onto next task.
+							 */
+							PopActiveSnapshot();
+							CommitTransactionCommand();
+							monitorExecutionState = ExecutionEnded;
```
Comment (marcocitus) [inline src/backend/distributed/utils/background_jobs.c:850]: the ExecutionEnded is confusing to me. This seems more like a TaskStillRunning state?

File: src/backend/distributed/utils/background_jobs.c (hunk 4 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:763]: we can actually open transaction at the start of the loop and commit it at the end to make it more readable. Thats what we do in the loop but unfriendly way.

File: src/backend/distributed/utils/background_jobs.c (hunk 5 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c]: I removed ExecutionEnded state, instead used readable flags representing loop breaking conditions.

File: src/backend/distributed/shared_library_init.c [hunk 6]
```diff
@@ -1689,6 +1689,24 @@ RegisterCitusConfigVariables(void)
...
+			"background tasks"),
+		gettext_noop(
+			"Controls the maximum number of parallel task executors the task monitor "
+			"can create for scheduled background tasks. Note that the value is not effective "
+			"if it is set a value higher than 'max_worker_processes' postgres parameter . It is "
+			"also not guaranteed to have exactly specified number of parallel task executors "
+			"because total background worker count is shared by all background workers. The value "
+			"represents the possible maximum number of task executors."),
+		&MaxBackgroundTaskExecutors,
+		4, 1, 8,
```
Comment (JelteF) [inline src/backend/distributed/shared_library_init.c:1705]: A maximum of 8 is way too low. And I think a default of 4 is also really on the low end. 
```suggestion
		4, 1, 1000,
```

File: src/test/regress/sql/background_task_queue_monitor.sql [hunk 7]
```diff
@@ -91,6 +91,44 @@ SELECT citus_job_wait(:job_id); -- wait for the job to be cancelled
...
+INSERT INTO pg_dist_background_task (job_id, command) VALUES (:job_id1, $job$ SELECT pg_sleep(5); $job$) RETURNING task_id AS task_id1 \gset
+INSERT INTO pg_dist_background_task (job_id, command) VALUES (:job_id1, $job$ SELECT pg_sleep(5); $job$) RETURNING task_id AS task_id2 \gset
+INSERT INTO pg_dist_background_task (job_id, command) VALUES (:job_id1, $job$ SELECT pg_sleep(5); $job$) RETURNING task_id AS task_id3 \gset
+INSERT INTO pg_dist_background_job (job_type, description) VALUES ('test_job', 'simple test to verify max parallel background execution') RETURNING job_id AS job_id2 \gset
+INSERT INTO pg_dist_background_task (job_id, command) VALUES (:job_id2, $job$ SELECT pg_sleep(5); $job$) RETURNING task_id AS task_id4 \gset
+INSERT INTO pg_dist_background_job (job_type, description) VALUES ('test_job', 'simple test to verify max parallel background execution') RETURNING job_id AS job_id3 \gset
+INSERT INTO pg_dist_background_task (job_id, command) VALUES (:job_id3, $job$ SELECT pg_sleep(5); $job$) RETURNING task_id AS task_id5 \gset
```
Comment (JelteF) [inline src/test/regress/sql/background_task_queue_monitor.sql:102]: Should be done in a transaction

File: src/backend/distributed/utils/background_jobs.c [hunk 8]
```diff
@@ -385,292 +648,344 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
-	}
+	ResetRunningBackgroundTasks();
 
+	PopActiveSnapshot();
+	CommitTransactionCommand();
 
-	MemoryContext oldContextPerJob = MemoryContextSwitchTo(perTaskContext);
-	TimestampTz backgroundWorkerFailedStartTime = 0;
+	/* create a map to store parallel task executors */
+	HTAB *backgroundExecutorHandles = CreateBackgroundExecutorHash(backgroundTaskContext);
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:664]: I recently made some nice helpers for hashtables that you can use instead now:
```suggestion
	HTAB *backgroundExecutorHandles = CreateSimpleHashWithName(int64, BackgroundExecutorHashEntry, "Background Executor Hash")`;
```

File: src/test/regress/sql/background_task_queue_monitor.sql [hunk 9]
```diff
@@ -91,6 +91,48 @@ SELECT citus_job_wait(:job_id); -- wait for the job to be cancelled
...
+    WHERE task_id IN (:task_id1, :task_id2, :task_id3, :task_id4, :task_id5)
+    ORDER BY job_id, task_id; -- show that last task is not running but ready to run(runnable)
+
+SELECT citus_job_cancel(:job_id2); -- when a job with 1 task is cancelled, the last runnable task will be running
+SELECT citus_job_wait(:job_id3, desired_status => 'running');
+SELECT job_id, task_id, status FROM pg_dist_background_task
+    WHERE task_id IN (:task_id1, :task_id2, :task_id3, :task_id4, :task_id5)
+    ORDER BY job_id, task_id;  -- show that last task is running
+
+SELECT citus_job_cancel(:job_id1);
```
Comment (JelteF) [inline src/test/regress/sql/background_task_queue_monitor.sql:118]: This tests something new as well: testing cancelling of multiple concurrently running tasks from the same job at the same time. Can you create some query to show that that works as expected? (i.e. `citus_job_wait`/`citus_task_wait` + `select ... from pg_dist_background_task ...`

File: src/backend/distributed/utils/background_jobs.c [hunk 10]
```diff
@@ -336,12 +567,21 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
+	MemoryContext backgroundTaskContext = AllocSetContextCreate(TopMemoryContext,
+																"BackgroundTaskContext",
+																ALLOCSET_DEFAULT_SIZES);
+
 	StartTransactionCommand();
 	PushActiveSnapshot(GetTransactionSnapshot());
 
-	/* load database name and copy to a memory context that survives the transaction */
+	/* make databasename alive during queue monitor lifetime */
+	MemoryContext oldContext = MemoryContextSwitchTo(backgroundTaskContext);
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:583]: Why did you move this up? It seems like that's unnecessary, any allocations that `get_database_name` would now also be part of the backgroundMemoryContext. But those shouldn't be necessary because we re-allocate databasename with `pstrdup`

File: src/backend/distributed/utils/background_jobs.c [hunk 11]
```diff
@@ -385,292 +625,347 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
-	while (hasTasks)
-	{
-		MemoryContextReset(perTaskContext);
+	QueueMonitorExecutionContext queueMonitorExecutionContext = {
+		.currentExecutorCount = 0,
+		.backgroundWorkerFailedStartTime = 0,
+		.ctx = backgroundTaskContext
+	};
 
+	/* loop exits if there is no running tasks left */
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:657]:
```suggestion
	/* loop exits if there is no running or runnable tasks left */
```

File: src/backend/distributed/utils/background_jobs.c [hunk 12]
```diff
@@ -310,6 +330,214 @@ CitusBackgroundTaskQueueMonitorErrorCallback(void *arg)
...
+
+	ereport(LOG, (errmsg("task jobid/taskid started: %ld/%ld",
+						 runnableTask->jobid, runnableTask->taskid)));
+
+	return true;
+}
+
+
+/*
+ * AssignRunnableTasks tries to assign all runnable tasks to a new task executor. If an assignment fails, it stops in case we hit some limitation.
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:495]: nit: optimal line length can be debated, but these comment lines are objectively quite long. Can you hard-wrap them to something reasonable (80/100/120 characters)?

(same for function above)

File: src/backend/distributed/utils/background_jobs.c [hunk 13]
```diff
@@ -385,292 +629,347 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
+						{
+							ereport(LOG, (errmsg("task jobid/taskid failed: %ld/%ld",
+												 task->jobid, task->taskid)));
+
+							monitorExecutionState = TaskHadError;
+						}
+						else if (mq_res == SHM_MQ_WOULD_BLOCK)
+						{
+							/*
+							 * still running the task but had no output for current iteration
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:836]: it's not necessarily true that it had no output right? It might have logged/noticed/warned stuff.

File: src/backend/distributed/utils/background_jobs.c [hunk 14]
```diff
@@ -385,292 +629,347 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
+
+						/* we are sure that at least one task did not block on current iteration */
+						allWouldBlock = false;
+
+						/*
+						 * we update task and job fields. We also update depending jobs.
+						 * At the end, do cleanup.
+						 */
+						UNSET_NULLABLE_FIELD(task, pid);
+						task->message = (handleEntry->message) ?
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:906]: when is handleEntry->message ever NULL?

File: src/backend/distributed/utils/background_jobs.c [hunk 15]
```diff
@@ -985,6 +1289,8 @@ StoreArgumentsInDSM(char *database, char *username, char *command, int64 taskId)
 	shm_toc_estimate_keys(&e, CITUS_BACKGROUND_TASK_NKEYS);
 	Size segsize = shm_toc_estimate(&e);
 
+	/* attaches the segment into TopMemoryContext, else, we can get garbage segment after commit */
+	CurrentResourceOwner = ResourceOwnerCreate(NULL, "citus background job");
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:1293]: Why did you move this CurrentResourceOwner assignment here from the CitusBackgroundTaskExecutor function? It's also missing the assignment to CurrentMemoryContext. Is that on purpose?

**Review by JelteF — CHANGES_REQUESTED** [review]

File: src/backend/distributed/utils/background_jobs.c [hunk 16]
```diff
@@ -385,292 +629,347 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
-			 * We still need to release the shared memory and xact before looping
+			 * starting from initial FSM state (ExecutionStarted), loop until
+			 * current task has no output (EWOULDBLOCK) or it is ended.
+			 * (suceeded, failed, or cancelled)
 			 */
+			while (!taskWouldBlock && !taskEnded)
+			{
+				switch (monitorExecutionState)
+				{
+					case ExecutionStarted:
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:715]: I don't think having this be a state machine adds anything to the readability. The flow seems completely linear:

1. ExecutionStarted
2. TaskConcurrentCancelCheck
3. TryConsumeTaskWorker
4. if (error) TaskHadError
5. if (ended) TaskEnded

File: src/backend/distributed/utils/background_jobs.c [hunk 17]
```diff
@@ -385,292 +629,347 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
+								backgroundWorkerFailedStartTime,
+								GetCurrentTimestamp(),
+								&secs, &microsecs);
+							ereport(LOG, (errmsg(
+											  "able to start a background worker with %ld seconds"
+											  "delay", secs)));
+
+							queueMonitorExecutionContext.backgroundWorkerFailedStartTime =
+								0;
+						}
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:743]: we need to reset the failure timestamp and also log time elapsed since last failure only after a successful worker allocation.

File: src/backend/distributed/utils/background_jobs.c [hunk 18]
```diff
@@ -310,6 +330,217 @@ CitusBackgroundTaskQueueMonitorErrorCallback(void *arg)
...
+								  MaxBackgroundTaskExecutors)));
+			queueMonitorExecutionContext->backgroundWorkerFailedStartTime =
+				GetCurrentTimestamp();
+		}
+
+		const long delay_ms = 1000;
+		(void) WaitLatch(MyLatch, WL_LATCH_SET | WL_TIMEOUT |
+						 WL_EXIT_ON_PM_DEATH,
+						 delay_ms, WAIT_EVENT_PG_SLEEP);
+		ResetLatch(MyLatch);
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:365]: We added sleep in case we cannot allocate worker, else the warning will be logged constantly if allWouldBlock = false.

File: src/backend/distributed/utils/background_jobs.c (hunk 15 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:1293]: It is deliberately NULL because "If CurrentResourceOwner is NULL, the segment remains attached until explicitly detached or the session ends."

File: src/backend/distributed/utils/background_jobs.c (hunk 13 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:836]: I removed the 'had no output' part but the shared memory queue has nothing to read if it returned SHM_MQ_WOULD_BLOCK.

File: src/backend/distributed/utils/background_jobs.c [hunk 19]
```diff
@@ -310,6 +335,422 @@ CitusBackgroundTaskQueueMonitorErrorCallback(void *arg)
...
+			queueMonitorExecutionContext->
+			backgroundWorkerFailedStartTime,
+			GetCurrentTimestamp(),
+			&secs, &microsecs);
+		ereport(LOG, (errmsg(
+						  "able to start a background worker with %ld seconds"
+						  "delay", secs)));
+
+		queueMonitorExecutionContext->backgroundWorkerFailedStartTime =
+			0;
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:565]:
```suggestion
		queueMonitorExecutionContext->backgroundWorkerFailedStartTime = 0;
```

File: src/backend/distributed/utils/background_jobs.c [hunk 20]
```diff
@@ -310,6 +335,422 @@ CitusBackgroundTaskQueueMonitorErrorCallback(void *arg)
...
+	char *databaseName = get_database_name(MyDatabaseId);
+	char *userName = GetUserNameFromId(runnableTask->owner, false);
+
+	/* try to create new executor and make it alive during queue monitor lifetime */
+	MemoryContext oldContext = MemoryContextSwitchTo(queueMonitorExecutionContext->ctx);
+	dsm_segment *seg = NULL;
+	BackgroundWorkerHandle *handle = StartCitusBackgroundTaskExecutor(databaseName,
+																	  userName,
+																	  runnableTask->
+																	  command,
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:448]: nit: breaking pointer dereferences like this across line isn't great to read. Sadly our autoformatting does that when there's not enough space on the line. If you put a newline between `StartCitusBackgroundTaskExecutor(` and `databaseName` it should be able to format it better.

File: src/backend/distributed/utils/background_jobs.c [hunk 21]
```diff
@@ -850,12 +1146,15 @@ UpdateDependingTasks(BackgroundTask *task)
...
 /*
- * ConsumeTaskWorkerOutput consumes the output of an executor and mutates the
- * BackgroundTask object to reflect changes like the message and status on the task.
+ * ConsumeTaskWorkerOutput consumes the output of an executor and sets the message as
+ * the last message read from the queue. It also sets hadError as true if executor had
+ * error.
  */
-static void
+static shm_mq_result
 ConsumeTaskWorkerOutput(shm_mq_handle *responseq, StringInfo message, bool *hadError)
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:1154]: To simplify calling this maybe we should return our own enum with three variants:
1. STILL_RUNNING/WOULDBLOCK/{some better name}
2. SUCCESS
3. ERROR

Then the hadError argument is not needed.

File: src/backend/distributed/utils/background_jobs.c [hunk 22]
```diff
@@ -385,292 +823,150 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
-			task->message = message.data;
-			UpdateBackgroundTask(task);
-			UpdateBackgroundJob(task->jobid);
+			/* create task execution context and assign it to queueMonitorExecutionContext */
+			TaskExecutionContext taskExecutionContext = {
+				.handleEntry = handleEntry,
+				.task = NULL,
+				.error = false
+			};
+			queueMonitorExecutionContext.taskExecutionContext = &taskExecutionContext;
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:898]: nit: I feel like conceptually this should be nested the other way around, i.e. queue monitor ctx should be inside a task execution context.

File: src/backend/distributed/utils/background_jobs.c [hunk 23]
```diff
@@ -385,292 +823,150 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
+			 * as we already consumed the queue.
 			 */
-			if (!task->retry_count)
-			{
-				SET_NULLABLE_FIELD(task, retry_count, 1);
-			}
-			else
+			BackgroundTaskStatus *taskExecutionStatus =
+				&(taskExecutionContext.task->status);
+			if (*taskExecutionStatus == BACKGROUND_TASK_STATUS_CANCELLED)
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:909]: nit: maybe return true or false from TaskConcurrentCancelCheck?

Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:909]: returned executionstatus enum from both TaskConcurrentCancelCheck and ConsumeTaskWorkerOutput.

File: src/backend/distributed/utils/background_jobs.c (hunk 22 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:898]: sounds better.

File: src/backend/distributed/utils/background_jobs.c [hunk 24]
```diff
@@ -986,6 +1288,14 @@ StoreArgumentsInDSM(char *database, char *username, char *command, int64 taskId)
 	Size segsize = shm_toc_estimate(&e);
 
 	dsm_segment *seg = dsm_create(segsize, DSM_CREATE_NULL_IF_MAXSEGMENTS);
+
+	/*
+	 * when we have CurrentResourceOwner != NULL, segment will be released upon CurrentResourceOwner release,
+	 * but we may consume the queue in segment even after CurrentResourceOwner released. 'dsm_pin_mapping' helps
+	 * persisting the segment until the session ends or the segment is detached explicitly by 'dsm_detach'.
+	 */
+	dsm_pin_mapping(seg);
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:1297]: Should be placed after the NULL check below

**Review by JelteF — APPROVED** [review]
